### DISCUSSION TITLE
How to Embed PostgreSQL (or other) databases in Tauri app for offline support

### DISCUSSION THREAD

**Discussion by hmShuvo314** [tauri-apps/tauri#5418] [Q&A] [answered]
Hello there,

I am building a Tauri app where I want to bring offline support. So for that, I would need a database (I am preferring PostgreSQL) embedded into my Tauri app so that the user can read and manipulate data even when there is no internet connection. Also, The application's users are non-technical people and they can not install PostgreSQL individually. So it has to be one single installer that bundles everything.

Any help of suggestion will be highly appreciated. Thanks in advance!

**Comment by FabianLars** [comment, marked as answer]
imo embedding _postgresql_ is not feasible at all and afaik not even possible. The only alternative would be to actually install postgresql on the user's system which sounds horrible to me.
So imho choosing a different database designed for this usecase, like sqlite or couchbase lite (nosql), is a more reasonable solution

**Reply by hmShuvo314** [reply]
Thanks for your reply. But our main database is built with PostgreSQL, and the local database in the Tauri app will need to be synchronized with the remote main database on the server once the app comes back online. So for the local database too, we need PostgreSQL.

I was reading about [Sidecar](https://tauri.app/v1/guides/building/sidecar/) in the Tauri docs. There they were talking about embedding NodeJS and Python binaries with Tauri app. But, can we also embed PostgreSQL like that? Kindly let me know.

**Reply by FabianLars** [reply]
> But, can we also embed PostgreSQL like that? Kindly let me know.

Nope, postgresql is not designed to be portable at all and to my knowledge there are no third party tools to make it portable other than this https://github.com/garethflowers/postgresql-portable which is Windows-only, doesn't look really trust-worthy to me, and the only person i know who tried to use it ended up switching to sqlite too, but i don't know why 🤷

**Reply by hmShuvo314** [reply]
Well, so if we can not, by any **viable** means, export the PostgreSQL database into a file (binary or other forms) and then upload or embed it with a Tauri app, could you please give me some directions regarding how I do the same with SQLite or MongoDB?

Also, our main remote server will still have PostgreSQL as the database, and we will need to synchronize the local database with the remote one. So do you think that will be a problem while managing to different databases (PostgreSQL on the main server, and maybe SQLite or MongoDB on the local client app). Thanks again for your time and help. I really appreciate it!

**Reply by FabianLars** [reply]
Sorryyy i totally forgot to answer.

tbh you didn't miss much because i have no idea how to sync the databases (i'm not a db guy) and for how to use sqlite etc:
If you don't want to use Rust then there is only this <https://github.com/tauri-apps/tauri-plugin-sql> (or using a sidecar).
If Rust is fine you can use sqlx, sqlite, etc for sqlite and for nosql dbs you could use [couchbase lite](https://crates.io/crates/couchbase-lite) for example. mongodb realm doesn't have rust bindings yet but it got a [c++ sdk](https://github.com/realm/realm-cpp) recently.

**Comment by dougg0k** [comment]
https://github.com/faokunega/pg-embed

https://github.com/fergusstrange/embedded-postgres

https://github.com/leinelissen/embedded-postgres

**Comment by cage1618** [comment]
sqlite3 is good choice

**Comment by diana96alazzam** [comment]
Can you share with me any updates regarding this issue? I have the same question but for MongoDB.

**Reply by hmShuvo314** [reply]
Hello, did this long time ago actually, and we ended up working with MariaDB instead. You can also embed MongoDB since it is designed to be portable as well. Lemme know if you need any further help

**Reply by jainsid499** [reply]
Hi I am building same kind of tauri desktop application, want to use sqlite to be embedded db and want to sync it with postgres in cloud. Is this possible?

**Reply by PizzaConsole** [reply]
@jainsid499 Yes it's possible

**Reply by jainsid499** [reply]
Hi, just wanted to clarify something, which ORM you used. Like I am using prisma , does it work perfectly well after bundling the app?

**Comment by PizzaConsole** [comment]
@hmShuvo314 Maybe look into this

https://supabase.com/blog/postgres-wasm

**Comment by aleclarson** [comment]
pglite seems promising

https://github.com/electric-sql/pglite

**Comment by SeanPedersen** [comment]
This is the way: https://github.com/theseus-rs/postgresql-embedded
